loris/utils.py

```diff
@@ -41,3 +42,31 @@ def symlink(src, dst, force=True):
         os.unlink(dst)
 
     os.symlink(src, dst)
+
+
+def safe_rename(src, dst):
+    """Rename a file from ``src`` to ``dst``.
+
+    We use a custom version rather than the standard library because we
+    have two requirements:
+
+    *   Moves must be atomic.  Otherwise Loris may serve a partial image from
+        a cache, which causes an error.  ``shutil.move()`` is not atomic.
+    *   Moves must work across filesystems.  Often temp directories and the
+        cache directories live on different filesystems.  ``os.rename()`` can
+        throw errors if run across filesystems.
+
+    So we try ``os.rename()``, but if we detect a cross-filesystem copy, we
+    switch to ``shutil.move()`` with some wrappers to make it atomic.
+    """
+    try:
+        os.rename(src, dst)
+    except OSError as err:
+        if err.errno == errno.EXDEV:
+            # First we move it across the filesystem to a temporary file
+            # in a non-atomic way, then move it to the final location
+            # atomically.
+            shutil.move(src, dst + '.tmp')
```

Comment:
As you noticed in another comment, using a unique generated name for the temporary file may be safer as different threads/processes could clash on the `.tmp` file. I don't think we need anything more than that because the end product of the different processes is the same content, and the only interaction would be on the `rename` calls, which are atomic.

Reply:
Yep, and the algorithm I posted (where did that comment go?) had provisions for multiple threads trying to do the copy.

Reply:
Looks like this is ready to merge once the tmp file has a unique file name, right?

Reply:
It’s a bit more complicated than that, but yes, implementing the whack-a-mole algorithm is the last thing to do.

Reply:
I think that both places where you're using safe_rename, the source is a tmp file, which can only have one thread/process writing to it. So I don't think a src getting deleted out from under a process is currently a problem we have to deal with.

If we do want to deal with that problem, how about adding a check before renaming dst-B.tmp to dst? How about something like this?

<pre>
if err.errno == errno.EXDEV:
    f = tempfile.NamedTemporaryFile(dir=..., delete=False)
    shutil.copyfile(src, f.name)
    if os.path.exists(src): #check that source file hasn't already been deleted by another process, so we know we got the whole file
        if not os.path.exists(dst):
            os.rename(f.name, dst)
            os.remove(src)
        else:
            pass #another process did the copy and will remove the src file
    else:
        pass #another process already did the copy and removed the src file
else:
    raise
</pre>